tests/behavior/transforms/remap.toml

```diff
@@ -1253,6 +1253,42 @@
         .a  && !.b  && .c && !.d  && .e && !.f ?? false
       '''
 
+[transforms.remap_function_tag_types_externally]
+  inputs = []
+  type = "remap"
+  source = """
+    .bytes = tag_types_externally("foo")
+    .integer = tag_types_externally(123)
+    .float = tag_types_externally(123.45)
+    .boolean = tag_types_externally(true)
+    .map = tag_types_externally({ "foo": "bar" })
+    .array = tag_types_externally(["foo"])
+    .timestamp = tag_types_externally(to_timestamp!("2021-01-01T00:00:00.00Z"))
+    .regex = tag_types_externally(/.*/)
+    .null = tag_types_externally(null)
+  """
+[[tests]]
+  name = "remap_function_tag_types_externally"
+  [tests.input]
+    insert_at = "remap_function_tag_types_externally"
+    type = "log"
+    log_fields = {}
+  [[tests.outputs]]
+    extract_from = "remap_function_tag_types_externally"
+    [[tests.outputs.conditions]]
+      type = "remap"
+      source = '''
+        .bytes == { "bytes": "foo" } && \
+        .integer == { "integer": 123 } && \
+        .float == { "float": 123.45 } && \
+        .boolean == { "boolean": true } && \
+        .map == { "foo": { "bytes": "bar" } } && \
+        .array == [{ "bytes": "foo" }] && \
+        .timestamp == { "timestamp": to_timestamp!("2021-01-01T00:00:00.00Z") } && \
+        .regex == { "regex": ".*" } && \
```

Comment:
This fails when I change it to

```
"regex": /.*/
```

Any ideas why?

Reply:
The CLI exposes the behavior I would have expected:

```
$ . = tag_types_externally(r'.*')
{ "regex": /.*/ }

$ .regex == r'.*'
true
```

Still not sure why this test only works with a string literal:

https://github.com/timberio/vector/blob/a497243d6abd8a248ef047f7a7144e4edcdf0e50/tests/behavior/transforms/remap.toml#L1074

Any idea, @JeanMertz?

Reply:
The CLI works entirely with `vrl::Value` types, which has a `VRL::Regex` type. When we interface with Vector, we convert VRL types to Vector types (`vector::Value`).

In Vector, we don't have a regex type because there is no sensible serialization strategy that really makes sense when going to JSON for example.

So, in your CLI example, you're comparing `vrl::Value` types, in the transform test, you're comparing `vector::Value` types, and we convert the regex type to a string representing the actual regex, which is why you're seeing what you're seeing.

At one point we actually returned an error when you tried to send a Regex type from VRL to Vector (basically meaning you could only assign regex types to variables, and couldn't have a type returning from your program that was or contained a regex). I removed this, as it was buggy and caused more trouble than it's worth for now, but we might re-introduce this if this is going to confuse people (I suspect it won't, but I can't be sure).

Reply:
One additional point.

This appears to be a bug:

```
$ . = tag_types_externally(r'.*')
{ "regex": /.*/ }
```

This should have said:

```
{ "regex": r'.*' }
```

Apparently, we haven't updated our `Display` impl for `Value` yet to switch from `/foo/` to `r'foo'`, so if you're willing to do so, go ahead (I'm okay with seeing that changed either in this PR, or a separate one).

Reply:
Opened https://github.com/timberio/vector/pull/6668.